### ISSUE TITLE
Verification fails (according to forward-auth logs) but forwarded to endpoint anyway with full _oauth code in URL

### ISSUE THREAD

**Issue by adawalli** [thomseddon/traefik-forward-auth#136]
Sorry for the long subject, but my error doesn't look like any other errors I have seen so far.

*  I am using Nomad for scheduling, but all of the "magic" comes through consul tags and therefore should look very familiar
* The user correctly gets redirected to azure for authentication  https://login.microsoftonline.com/<tenant_id>/v2.0
* The user gets redirected back to the endpoint URL (`https://<whoami-service>.<domain>/_oauth`
* However - the middleware kicks this up the chain, and the user ends up at the protected endpoint with the full _oauth string in the URL

This is the Consul service stanza that creates the `forward-auth` middleware

```hcl
service {
      name = "forward-auth"
      port = "auth"
      tags = [
        "traefik.enable=true",
        "traefik.http.middlewares.forward-auth.forwardauth.address=http://localhost:4181",
        "traefik.http.middlewares.forward-auth.forwardauth.authResponseHeaders=X-Forwarded-User",
        "traefik.http.routers.forward-auth.middlewares=forward-auth@consulcatalog"
        ]
    }
```

It's important to note that the Forward-Auth port (of 4181) is bound to localhost into the network namespace for `traefik` - that's why http://localhost:4181 works

Here is the config for the "Forward-Auth" container

```hcl
task "forward-auth" {
      driver = "docker"
      config {
        image = "thomseddon/traefik-forward-auth:2"
      }
      env {
        DEFAULT_PROVIDER = "oidc"
        PROVIDERS_OIDC_ISSUER_URL = "https://login.microsoftonline.com/<redacted_tenant>/v2.0"
        PROVIDERS_OIDC_CLIENT_ID = "<redacted_client_id>"
        PROVIDERS_OIDC_CLIENT_SECRET = "<redacted_client_secret>"
        SECRET = "<base64_encoded_secret>"
        COOKIE_DOMAIN = "<redacted domain>"
        LOG_LEVEL = "trace"
        LOG_FORMAT = "pretty"
      }
    }
  }
```

Here is the `whoami` container that I am trying to **protect** for the sake of demonstration
```hcl
 task "whoami" {
      driver = "docker"
      config {
        image = "containous/whoami"
        port_map {
          http = 80
        }
      }
      resources {
        network {
          port "http" {}
        }
      }

      service {
        port = "http"
        tags = [
          "traefik.http.routers.whoami.rule=Host(`<whoami-service>.<redacted domain>`)",
          "traefik.enable=true",
          "traefik.http.routers.whoami.tls=true",
          "traefik.http.routers.whoami.entryPoints=webSecure",
          "traefik.http.routers.whoami.middlewares=forward-auth@consulcatalog"
        ]
      }
    }
```

When the authentication fails, and I get dumped at at the full URL with `/_oauth?code=` in tow, I see the following log in the forward-auth container
```
time="2020-06-26T21:03:21Z" level=error msg="Error getting user: failed to verify signature: failed to verify id token signature" source_ip=10.0.2.2
```

Please see a more complete log below
```
time="2020-06-26T21:03:15Z" level=debug msg="Authenticating request" headers="map[Accept:[text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9] Accept-Encoding:[gzip, deflate, br] Accept-Language:[en-US,en;q=0.9] Cookie:[_forward_auth_csrf=97ef7d6a1eb85e16391931fb418ce4ae] Dnt:[1] Sec-Fetch-Dest:[document] Sec-Fetch-Mode:[navigate] Sec-Fetch-Site:[none] Sec-Fetch-User:[?1] Upgrade-Insecure-Requests:[1] User-Agent:[Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.106 Safari/537.36] X-Forwarded-For:[10.0.2.2] X-Forwarded-Host:[<whoami-service>.<redacted domain>:8082] X-Forwarded-Method:[GET] X-Forwarded-Port:[8082] X-Forwarded-Proto:[https] X-Forwarded-Server:[node1] X-Forwarded-Uri:[/] X-Real-Ip:[10.0.2.2]]" rule=default source_ip=10.0.2.2
time="2020-06-26T21:03:15Z" level=debug msg="Set CSRF cookie and redirecting to google login" source_ip=10.0.2.2
time="2020-06-26T21:03:15Z" level=debug msg=Done source_ip=10.0.2.2
time="2020-06-26T21:03:15Z" level=debug msg="Handling callback" headers="map[Accept:[text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9] Accept-Encoding:[gzip, deflate, br] Accept-Language:[en-US,en;q=0.9] Cookie:[_forward_auth_csrf=dd540334ddc26a93513e485f81a73dde] Dnt:[1] Sec-Fetch-Dest:[document] Sec-Fetch-Mode:[navigate] Sec-Fetch-Site:[none] Sec-Fetch-User:[?1] Upgrade-Insecure-Requests:[1] User-Agent:[Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.106 Safari/537.36] X-Forwarded-For:[10.0.2.2] X-Forwarded-Host:[<whoami-service>.<redacted domain>:8082] X-Forwarded-Method:[GET] X-Forwarded-Port:[8082] X-Forwarded-Proto:[https] X-Forwarded-Server:[node1] X-Forwarded-Uri:[/_oauth?code=<redacted_code>&state=dd540334ddc26a93513e485f81a73dde%3aoidc%3ahttps%3a%2f%2f<service>.<redacted domain>%3a8082%2f&session_state=b3626b37-b0d8-4256-a229-16b575833ce4] X-Real-Ip:[10.0.2.2]]" rule=default source_ip=10.0.2.2
time="2020-06-26T21:03:21Z" level=error msg="Error getting user: failed to verify signature: failed to verify id token signature" source_ip=10.0.2.2
```

Also, the `"Set CSRF cookie and redirecting to google login" ` string is interesting...considering I am using Azure for my OIDC

Thanks in advance - this is a really cool project

**Comment by adawalli** [comment]
FWIW, I tried exposing forward-auth as a service through Host rule, but I get exactly the same results

```hcl
    service {
      name = "forward-auth"
      port = "auth"
      tags = [
        "traefik.enable=true",
        "traefik.http.routers.forward-auth.entrypoints=web",
        "traefik.http.routers.forward-auth.rule=Host(`auth.<redacted-domain>`)",
        "traefik.http.middlewares.forward-auth.forwardauth.address=http://auth.<redacted-domain>:4181",
        "traefik.http.middlewares.forward-auth.forwardauth.authResponseHeaders=X-Forwarded-User",
        "traefik.http.routers.forward-auth.middlewares=forward-auth@consulcatalog",
        "traefik.http.services.forward-auth.loadbalancer.server.port=4181"
        ]
    }
```

**Comment by thomseddon** [comment]
Hmm, the service instructs the authorisation server to return the client to `/_oauth` (i.e. that's the redirect_uri), which would explain why you end up there.

The "failed to verify signature: failed to verify id token signature" is actually being kicked out by go-oidc from here: https://github.com/thomseddon/traefik-forward-auth/blob/8b3a950162a745ba9dcda62fa1a8dceb3c7464cf/internal/provider/oidc.go#L87-L91

Were the client credentials generated by registering an app via Azure Active Directory or by some other means? When googling, I can see a few issues relating to the way the credentials were created, apparently graph credentials will not work?

**Comment by thomseddon** [comment]
I've added some extra details on how to setup and use Microsoft as a provider: https://github.com/thomseddon/traefik-forward-auth/wiki/Provider-Setup#microsoft

**Comment by adawalli** [comment]
@thomseddon - this was through a registered app on Azure. I will take a look at your instructions to see if a step was missed somehow

**Comment by adawalli** [comment]
I followed your steps to a T and I still have the same result. A few questions

* If the redirect is being followed correctly, to _oauth, and my token fails...shouldn't I at least get kicked out? Instead, I get to see the protected resource (albeit with the _oauth string in the URL)
* Is there some scope or permission I should be requesting, but might not be currently? Currently, my azure app has: `email`, `offline_access`, `openid`, `profile`, and `User.Read` - FWIW, these are all of type "Delegated"

**Comment by thomseddon** [comment]
Yes, I actually discovered a bug yesterday which I fixed in #142 (and added some extra tests to ensure it's not reintroduced). This was released in v2.2 yesterday - could you try with that?

I think you will still see an error, I will try and test with the Microsoft provider this week to make sure the docs are correct.

Thanks for the feedback

**Comment by adawalli** [comment]
The new image definitely seems to be an improvement, although not quite what I expect.

The flow happens like this

* Forwarded to login.microsoft as expected
* Redirected back to http://myapp/_oauth
* I now see "service unavailable" which is good - but I still see the whole _oauth string in the URL, which is weird
* Interestingly, if I hit "Refresh", I see "Not Authorized" (but _oauth still in the URL)

I see a couple of folks [here](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/609) talk about forcing a JWT request by creating an API scope and adding that to the scope request. Is there a way I can request a scope when using OIDC as a provider?

**Comment by adawalli** [comment]
also @thomseddon , I am on keybase if you want to setup a chat. I would love to help you debug this and root-cause the issue!

**Comment by thomseddon** [comment]
I just tested the microsoft login and it worked nicely once I got my tenant/client IDs the right way around (I've updated the provider setup docs).

I just created a brand new app and didn't have to grant it any specific permissions (it only seems to have the User.Read permission, also showing as Delegated).

Is it still failing with the same log message?

**Comment by adawalli** [comment]
Double checked my setup, and it's exactly as you said - I will try to use your docker-compose example just to rule out any weird nomad issues.

**Comment by adawalli** [comment]
I tried your exact OIDC Example for docker swarm, and see exactly the same issue. Here are a few screenshots of my Azure setup

<img width="1130" alt="api_perms" src="https://user-images.githubusercontent.com/14079121/86298567-0d5ad300-bbcc-11ea-8e79-9e44aca9f12a.png">

<img width="1063" alt="certs_secrets" src="https://user-images.githubusercontent.com/14079121/86298579-121f8700-bbcc-11ea-98a0-a51a5204e4a6.png">

<img width="845" alt="grant" src="https://user-images.githubusercontent.com/14079121/86298591-15b30e00-bbcc-11ea-8e61-3e77cdfd061d.png">

<img width="1106" alt="redirect" src="https://user-images.githubusercontent.com/14079121/86298603-1a77c200-bbcc-11ea-995c-68c33b6fc08b.png">

Does anything seem out of place?

I am getting 

```
traefik-forward-auth_1  | time="2020-07-01T22:48:03Z" level=debug msg="Handling callback" cookies="[session=c2643140-ae9e-4374-8d38-a495a2757fef _forward_auth_csrf=685721109277ba645c5cf627ff8c6867]" handler=AuthCallback host="localhost:8085" method=GET proto=http rule=default source_ip=172.18.0.1 uri="/_oauth?code=OAQABAAIAAAAGV_bv21oQQ4ROqh0_1-tApqH85li9AL5nM3dij79aRj1jjHh5kVwIwLk-xeJtj8wqsPyZxoDRrPRDumAZd8eCx7sexV5656Z3FsdTK3CHW6d7-pMPEvhFKmLvPcwlK05O0O-E_Dqru0hc6uegD4c0Ha9_pKjngPMxZVDk_ibPiBKMEI27klBb0JGswiHjdFjHViR8QCeiAgtD_1a8fLVpP_vletg1wXfoMozw-JdfMHjxjOaESRcM5k77w5-DPaKPiZuiwz7wSv26Rsg4DTxHlr9vyDNRdHQIlCIGovWX_1zbSByIhccwCuh0Rs2-MdPg2hrDABxz-qm7c9X5dZIuUw3rDMmYI1AJDWRC1jeNXPQCcpghHaGXzQcz4zzXYzhIAKXmVqXwHW-csODIyfO86EJAxRtAp1FN_QKfERpl5ves9EvO7Tihhee4npgPoMpiPpGDGpI4Sn0EBsLzp1Wrq1RkRig6OTM-TFb0KDpFCFyUD2pmmTy9sTn99Wxv5ub0--N8gqXtBpwjMaWivSHJxZ5YvDas-k2HurOUlFRp48jqscHQOj80Hwe_F7AABrnSXl0CmXbhs2uX1HXWZviGNqXW8fa51AhtzOAW_rPjYYHEpiHpSYj-pEvogXwVvyDlJTR8cOthughhTVDraWKmxFbkD7g1sobUZkyN_HuW3W5rzVmE68Ymh56TUJFuVLp6wfVYrtXKaByIsLEZZOXr3BvSVUsZ1AJoDeiL2M-1FyAA&state=685721109277ba645c5cf627ff8c6867%3aoidc%3ahttp%3a%2f%2flocalhost%3a8085%2ffavicon.ico&session_state=a9cfc5ff-bad9-4061-bfb0-764722d1a94c"
traefik-forward-auth_1  | time="2020-07-01T22:48:04Z" level=error msg="Error getting user" error="failed to verify signature: failed to verify id token signature" handler=AuthCallback host="localhost:8085" method=GET proto=http rule=default source_ip=172.18.0.1 uri="/_oauth?code=OAQABAAIAAAAGV_bv21oQQ4ROqh0_1-tApqH85li9AL5nM3dij79aRj1jjHh5kVwIwLk-xeJtj8wqsPyZxoDRrPRDumAZd8eCx7sexV5656Z3FsdTK3CHW6d7-pMPEvhFKmLvPcwlK05O0O-E_Dqru0hc6uegD4c0Ha9_pKjngPMxZVDk_ibPiBKMEI27klBb0JGswiHjdFjHViR8QCeiAgtD_1a8fLVpP_vletg1wXfoMozw-JdfMHjxjOaESRcM5k77w5-DPaKPiZuiwz7wSv26Rsg4DTxHlr9vyDNRdHQIlCIGovWX_1zbSByIhccwCuh0Rs2-MdPg2hrDABxz-qm7c9X5dZIuUw3rDMmYI1AJDWRC1jeNXPQCcpghHaGXzQcz4zzXYzhIAKXmVqXwHW-csODIyfO86EJAxRtAp1FN_QKfERpl5ves9EvO7Tihhee4npgPoMpiPpGDGpI4Sn0EBsLzp1Wrq1RkRig6OTM-TFb0KDpFCFyUD2pmmTy9sTn99Wxv5ub0--N8gqXtBpwjMaWivSHJxZ5YvDas-k2HurOUlFRp48jqscHQOj80Hwe_F7AABrnSXl0CmXbhs2uX1HXWZviGNqXW8fa51AhtzOAW_rPjYYHEpiHpSYj-pEvogXwVvyDlJTR8cOthughhTVDraWKmxFbkD7g1sobUZkyN_HuW3W5rzVmE68Ymh56TUJFuVLp6wfVYrtXKaByIsLEZZOXr3BvSVUsZ1AJoDeiL2M-1FyAA&state=685721109277ba645c5cf627ff8c6867%3aoidc%3ahttp%3a%2f%2flocalhost%3a8085%2ffavicon.ico&session_state=a9cfc5ff-bad9-4061-bfb0-764722d1a94c"
```

**Comment by adawalli** [comment]
ok, got it to work finally. I went into the API menu and deleted all of the scopes from there and now it's all working!

**Comment by adawalli** [comment]
Closing this issue - I think that the enterprise app had too many permissions - when I leave it set to the stock permissions, OIDC Login + ForwardAuth is working as documented

**Comment by benjyiw** [comment]
I was having the same issue that was described here. Wanted to include some stuff that I found in order to resolve it in my  scenario.

When you create an Azure App Registration in the Web Portal, Azure automatically adds `required_resource_access` for `User.Read`. This is the only permission required to get it to work. In my case, I was using Terraform to create the Azure App Registration, which did not create the required_resource_access automatically. In order to resolve that with Terraform I had to include that permission (which is a bit cryptic consider microsofts use of guids to specify user.read):

```
resource "azuread_application" "app" {
  name                       = "myapp"
  reply_urls                 = "https://whatever.example.com/_oauth"
  type                       = "webapp/api"
  available_to_other_tenants = "false"

  required_resource_access {
    resource_app_id = "00000003-0000-0000-c000-000000000000"

    resource_access {
      id   = "e1fe6dd8-ba31-4d61-89e7-88639da4683d" # User.Read
      type = "Scope"
    }
  }
}
```

The same would apply if you were automating the app registration with other infrastructure management tools or powershell or whatever.
